platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultConfigurationCache.kt

```diff
@@ -369,11 +394,15 @@ class DefaultConfigurationCache internal constructor(
     }
 
     private
-    fun determineCacheAction(): Pair<ConfigurationCacheAction, StructuredMessage> = when {
+    fun determineCacheAction(): DescribedAction =
+        determineBasicCacheAction().downgradeIfNeeded()
```

Comment:
💭  This two-step determination is still looking confusing to me. This is not a blocker to me, and I see it's a matter of  taste. I'd prefer to encapsulate this decision making during cache action determining, where for each store/update `when` branch returns something like `Store orSkipStore to description `

Reply:
I tried to make it explicit that figuring out the action is a two-level process - there is the regular/basic action (computed via `determineBasicCacheAction()`), and in the rare case we are in read-only mode, we should replace it with `SkipStore` if the action is not read-only itself. I changed `determineCacheAction()` to:

```
    private
    fun determineCacheAction(): DescribedAction {
        val basicAction = determineBasicCacheAction()
        if (startParameter.isReadOnlyCache && !basicAction.action.isReadOnly) {
            // fall back to [SkipStore] if we are in read-only mode
            return SkipStore.withDescription(basicAction.description)
        }
        return basicAction
    }
```

I hope the two-level determination should be explicit enough now.